flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -659,14 +661,34 @@ private CompletableFuture<CleanupJobState> handleJobManagerRunnerResult(
         return jobReachedTerminalState(jobManagerRunnerResult.getExecutionGraphInfo());
     }
 
-    enum CleanupJobState {
-        LOCAL,
-        GLOBAL
+    private static class CleanupJobStateWithJobStatusIfGlobal {
+
+        private static final CleanupJobStateWithJobStatusIfGlobal LOCAL = new CleanupJobStateWithJobStatusIfGlobal(null);
+
+        @Nullable private final JobStatus jobStatus;
+
+        public static CleanupJobStateWithJobStatusIfGlobal globalCleanup(JobStatus jobStatus) {
+            return new CleanupJobStateWithJobStatusIfGlobal(jobStatus);
+        }
+
+        private CleanupJobStateWithJobStatusIfGlobal(@Nullable JobStatus jobStatus) {
+            this.jobStatus = jobStatus;
+        }
+
+        public boolean isGlobalCleanup() {
+            return jobStatus != null;
+        }
+
+        public JobStatus getJobStatus() {
+            Preconditions.checkState(
```

Comment:
I'm not really happy about this Precondition. But I couldn't come up with a better way to pass in the JobStatus without touching other method signatures. Do you have a better idea?

Reply:
Not really - either the precondition check happens here or it happens at the `getJobStatus` callsites. Other than a checked exception, I can't see a nice way of doing it.